keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1328 @@
++
++	// The individual named trust anchor set within the ClusterTrustBundle.
++	TrustAnchorSetName string `json:"trustAnchorSetName"`
++}
+```
+
+### trustAnchors Projected Volume Source
+
+The trustAnchors projected volume source is responsible for retrieving a
+designated trust anchor set from a signer and publishing it into the container
+filesystem.
+
+All of the new fields described in this section are gated by the
+ClusterTrustBundle kubelet feature gate.  Unless the feature gate is enabled,
+usage of these alpha objects and fields will be rejected by kubelet.
+
+#### Configuration Object Definition
+
+```go
+// TrustAnchorsProjection is the top-level configuration object for
+// instances of the trustAnchors kubelet projected volume source.
+type TrustAnchorsProjection struct {
+  // Which trust anchor sets should we retrieve from the signer, and where
+  // should we write them into the volume?
+  Files []TrustAnchorSetFile `json:"files"`
+}
+
+// TrustAnchorsFile is a oneof that holds different strategies for writing
+// trust anchor sets to the projected volume.
+//
+// Conceptually, it’s a oneof, but only the FlatFile option is provided
+// for now.  In the future, different formats that can hold more than one
+// named trust anchor set may be supported.  Or we might support specialty
+// formats like Java keystores.
+type TrustAnchorsFile struct {
+  // If set, write a single trust anchor set to a single PEM file.
+  FlatFile *FlatFile `json:"flatFile"`
+}
+
+// FlatFile is a trust anchor set provisioning strategy that writes a single
+// named trust anchor set to a single file within the project volume.
+//
+// The file format is newline-separated PEM-wrapped DER-encoded X.509
+// certificates.
+type FlatFile struct {
+  // An explicit ClusterTrustBundle to retrieve trust anchor sets from.
+  ClusterTrustBundleName string `json:"clusterTrustBundleName"`
+
+  // Which named trust anchor set should we read from the ClusterTrustBundle
+  // object?
+  SetName string `json:"setName"`
+
+  // Where should we write the flat file in the projected volume?
+  Path string `json:"path"`
+}
+```
+
+#### Volume Content Generation and Refresh
+
+For each flatFile entry in each trustAnchors volume source in the pod spec:
+
+1. Kubelet will establish a watch on the named ClusterTrustBundle (or use a
+   pre-existing global watch).
+
+2. Kubelet will wait for all relevant ClusterTrustBundle objects to have
+   the "Ready" condition.
+
+3. Kubelet will retrieve the named set from the ClusterTrustBundle object,
+   and write it into the projected volume at path.
+
+The volume will not be successfully mounted (and thus the pod will not start)
+until all certificate files have been written to the container filesystem.
+
+Any errors in this process will be surfaced as a volume mount errors that block
+the pod from starting up, since it's expected that the pod will likely crash or
+misbehave if the certificates it needs are not present.
```

Comment:
Will events be generated?

Reply:
I don't really have enough experience to know what the correct answer here is.  I know that when I implemented a similar system using CSI drivers, errors returned from the CSI driver showed up when I did `kubectl describe pod <mypod>`.  I guess that's because Kubelet is emitting events?  If so, then yes, we should do the same thing.

Reply:
I think we need something for pod authors to inspect and know why their pod isn't starting